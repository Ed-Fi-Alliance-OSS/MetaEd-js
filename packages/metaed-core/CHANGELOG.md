# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.1.1-dev.35"></a>
## [1.1.1-dev.35](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.34...v1.1.1-dev.35) (2018-02-14)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.34"></a>
## [1.1.1-dev.34](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.32...v1.1.1-dev.34) (2018-01-18)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.33"></a>
## [1.1.1-dev.33](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.32...v1.1.1-dev.33) (2018-01-18)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.32"></a>
## [1.1.1-dev.32](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.31...v1.1.1-dev.32) (2018-01-17)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.31"></a>
## [1.1.1-dev.31](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.30...v1.1.1-dev.31) (2018-01-15)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.30"></a>
## [1.1.1-dev.30](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.29...v1.1.1-dev.30) (2018-01-15)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.29"></a>
## [1.1.1-dev.29](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.26...v1.1.1-dev.29) (2018-01-14)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.28"></a>
## [1.1.1-dev.28](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.26...v1.1.1-dev.28) (2017-12-20)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.27"></a>
## [1.1.1-dev.27](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.26...v1.1.1-dev.27) (2017-12-20)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.26"></a>
## [1.1.1-dev.26](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.25...v1.1.1-dev.26) (2017-12-18)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.25"></a>
## [1.1.1-dev.25](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.23...v1.1.1-dev.25) (2017-12-15)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.24"></a>
## [1.1.1-dev.24](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.23...v1.1.1-dev.24) (2017-12-15)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.23"></a>
## [1.1.1-dev.23](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.22...v1.1.1-dev.23) (2017-12-13)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.22"></a>
## [1.1.1-dev.22](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.18...v1.1.1-dev.22) (2017-12-13)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.20"></a>
## [1.1.1-dev.20](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.18...v1.1.1-dev.20) (2017-12-13)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.19"></a>
## [1.1.1-dev.19](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.18...v1.1.1-dev.19) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.18"></a>
## [1.1.1-dev.18](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.16...v1.1.1-dev.18) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.17"></a>
## [1.1.1-dev.17](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.16...v1.1.1-dev.17) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.16"></a>
## [1.1.1-dev.16](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.15...v1.1.1-dev.16) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.15"></a>
## [1.1.1-dev.15](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.14...v1.1.1-dev.15) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.14"></a>
## [1.1.1-dev.14](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.13...v1.1.1-dev.14) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.13"></a>
## [1.1.1-dev.13](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.12...v1.1.1-dev.13) (2017-12-11)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.12"></a>
## [1.1.1-dev.12](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.11...v1.1.1-dev.12) (2017-12-10)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.11"></a>
## [1.1.1-dev.11](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.10...v1.1.1-dev.11) (2017-12-09)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.10"></a>
## [1.1.1-dev.10](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.9...v1.1.1-dev.10) (2017-12-08)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.9"></a>
## [1.1.1-dev.9](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.8...v1.1.1-dev.9) (2017-12-08)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.8"></a>
## [1.1.1-dev.8](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.7...v1.1.1-dev.8) (2017-12-08)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.7"></a>
## [1.1.1-dev.7](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.6...v1.1.1-dev.7) (2017-12-08)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.6"></a>
## [1.1.1-dev.6](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.5...v1.1.1-dev.6) (2017-12-04)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.5"></a>
## [1.1.1-dev.5](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.4...v1.1.1-dev.5) (2017-11-29)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.4"></a>
## [1.1.1-dev.4](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.3...v1.1.1-dev.4) (2017-11-27)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.2"></a>
## [1.1.1-dev.2](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.1...v1.1.1-dev.2) (2017-11-27)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.1"></a>
## [1.1.1-dev.1](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.1.1-dev.0...v1.1.1-dev.1) (2017-11-27)




**Note:** Version bump only for package metaed-core

<a name="1.1.1-dev.0"></a>
## [1.1.1-dev.0](https://github.com/Ed-Fi-Alliance/MetaEd-js/compare/v1.0.0-dev.5...v1.1.1-dev.0) (2017-11-21)




**Note:** Version bump only for package metaed-core
