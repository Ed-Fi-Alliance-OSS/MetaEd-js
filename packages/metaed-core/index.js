// @flow

// Test helper -- *** where should this go? ***
export { MetaEdTextBuilder } from './test/MetaEdTextBuilder';


// Functions

export { deepFreeze, deepFreezeAssign, prependIndefiniteArticle } from './src/Utility';
export { startingFromFileLoadP } from './src/task/Pipeline';
export { createMetaEdFile } from './src/task/MetaEdFile';
export { loadCoreBufferedFiles, loadExtensionBufferedFiles } from './src/task/BufferFileLoader';
export { loadFiles } from './src/task/FileSystemFilenameLoader';
export { newState } from './src/State';
export { addEntity, getAllEntities, getAllTopLevelEntities, getEntity, getTopLevelCoreEntity, getEntitiesOfType, getAllEntitiesNoSimpleTypes } from './src/model/EntityRepository';
export { addProperty, getAllProperties, getPropertiesOfType, newPropertyIndex } from './src/model/property/PropertyRepository';
export { isReferenceProperty, isSharedProperty, allPropertyTypes } from './src/model/property/PropertyType';
export { allEntityModelTypes, allTopLevelEntityModelTypes, topLevelCoreEntityModelTypes } from './src/model/ModelType';

// Builders

export { AssociationBuilder } from './src/builder/AssociationBuilder';
export { AssociationExtensionBuilder } from './src/builder/AssociationExtensionBuilder';
export { AssociationSubclassBuilder } from './src/builder/AssociationSubclassBuilder';
export { ChoiceBuilder } from './src/builder/ChoiceBuilder';
export { CommonBuilder } from './src/builder/CommonBuilder';
export { CommonExtensionBuilder } from './src/builder/CommonExtensionBuilder';
export { DecimalTypeBuilder } from './src/builder/DecimalTypeBuilder';
export { DescriptorBuilder } from './src/builder/DescriptorBuilder';
export { DomainBuilder } from './src/builder/DomainBuilder';
export { DomainEntityBuilder } from './src/builder/DomainEntityBuilder';
export { DomainEntityExtensionBuilder } from './src/builder/DomainEntityExtensionBuilder';
export { DomainEntitySubclassBuilder } from './src/builder/DomainEntitySubclassBuilder';
export { EnumerationBuilder } from './src/builder/EnumerationBuilder';
export { IntegerTypeBuilder } from './src/builder/IntegerTypeBuilder';
export { InterchangeBuilder } from './src/builder/InterchangeBuilder';
export { NamespaceInfoBuilder } from './src/builder/NamespaceInfoBuilder';
export { SharedDecimalBuilder } from './src/builder/SharedDecimalBuilder';
export { SharedIntegerBuilder } from './src/builder/SharedIntegerBuilder';
export { SharedStringBuilder } from './src/builder/SharedStringBuilder';
export { StringTypeBuilder } from './src/builder/StringTypeBuilder';

// Flow types

export type { State } from './src/State';
export type { InputDirectory } from './src/task/FileSystemFilenameLoader';
export type { FileSet, MetaEdFile } from './src/task/MetaEdFile';
export type { FileIndex } from './src/task/FileIndex';
export type { Validator } from './src/validator/Validator';
export type { ValidationFailure } from './src/validator/ValidationFailure';
export type { Enhancer } from './src/enhancer/Enhancer';
export type { EnhancerResult } from './src/enhancer/EnhancerResult';
export type { Generator } from './src/generator/Generator';
export type { GeneratorResult } from './src/generator/GeneratorResult';
export type { GeneratedOutput } from './src/generator/GeneratedOutput';
export type { MetaEdEnvironment, PluginEnvironment, SemVer } from './src/MetaEdEnvironment';
export type { MetaEdPlugin, PluginManifest } from './src/plugin/PluginTypes';

// Properties
export type { AssociationProperty } from './src/model/property/AssociationProperty';
export type { BooleanProperty } from './src/model/property/BooleanProperty';
export type { ChoiceProperty } from './src/model/property/ChoiceProperty';
export type { CommonProperty } from './src/model/property/CommonProperty';
export type { CurrencyProperty } from './src/model/property/CurrencyProperty';
export type { DateProperty } from './src/model/property/DateProperty';
export type { DecimalProperty } from './src/model/property/DecimalProperty';
export type { DescriptorProperty } from './src/model/property/DescriptorProperty';
export type { DomainEntityProperty } from './src/model/property/DomainEntityProperty';
export type { DurationProperty } from './src/model/property/DurationProperty';
export type { EntityProperty, HasReferencedEntity } from './src/model/property/EntityProperty';
export type { EnumerationProperty } from './src/model/property/EnumerationProperty';
export type { InlineCommonProperty } from './src/model/property/InlineCommonProperty';
export type { IntegerProperty } from './src/model/property/IntegerProperty';
export type { MergedProperty } from './src/model/property/MergedProperty';
export type { PercentProperty } from './src/model/property/PercentProperty';
export type { PropertyIndex } from './src/model/property/PropertyRepository';
export type { PropertyType } from './src/model/property/PropertyType';
export type { ReferentialProperty } from './src/model/property/ReferentialProperty';
export type { SchoolYearEnumerationProperty } from './src/model/property/SchoolYearEnumerationProperty';
export type { SharedDecimalProperty } from './src/model/property/SharedDecimalProperty';
export type { SharedIntegerProperty } from './src/model/property/SharedIntegerProperty';
export type { SharedShortProperty } from './src/model/property/SharedShortProperty';
export type { SharedStringProperty } from './src/model/property/SharedStringProperty';
export type { ShortProperty } from './src/model/property/ShortProperty';
export type { SimpleProperty } from './src/model/property/SimpleProperty';
export type { SharedSimple } from './src/model/SharedSimple';
export type { StringProperty } from './src/model/property/StringProperty';
export type { TimeProperty } from './src/model/property/TimeProperty';
export type { YearProperty } from './src/model/property/YearProperty';

// Entities
export type { AbstractEntity } from './src/model/AbstractEntity';
export type { Association } from './src/model/Association';
export type { AssociationExtension } from './src/model/AssociationExtension';
export type { AssociationSubclass } from './src/model/AssociationSubclass';
export type { Choice } from './src/model/Choice';
export type { Common } from './src/model/Common';
export type { CommonExtension } from './src/model/CommonExtension';
export type { DecimalType } from './src/model/DecimalType';
export type { Descriptor } from './src/model/Descriptor';
export type { Domain } from './src/model/Domain';
export type { DomainEntity } from './src/model/DomainEntity';
export type { DomainEntityExtension } from './src/model/DomainEntityExtension';
export type { DomainEntitySubclass } from './src/model/DomainEntitySubclass';
export type { DomainItem } from './src/model/DomainItem';
export type { EntityRepository } from './src/model/EntityRepository';
export type { Enumeration } from './src/model/Enumeration';
export type { EnumerationItem } from './src/model/EnumerationItem';
export type { IntegerType } from './src/model/IntegerType';
export type { Interchange } from './src/model/Interchange';
export type { InterchangeExtension } from './src/model/InterchangeExtension';
export type { InterchangeItem } from './src/model/InterchangeItem';
export type { MapTypeEnumeration } from './src/model/MapTypeEnumeration';
export type { ModelBase } from './src/model/ModelBase';
export type { ModelType } from './src/model/ModelType';
export type { NamespaceInfo } from './src/model/NamespaceInfo';
export type { SchoolYearEnumeration } from './src/model/SchoolYearEnumeration';
export type { SharedDecimal } from './src/model/SharedDecimal';
export type { SharedInteger } from './src/model/SharedInteger';
export type { SharedString } from './src/model/SharedString';
export type { SourceMap } from './src/model/SourceMap';
export type { StringType } from './src/model/StringType';
export type { Subdomain } from './src/model/Subdomain';
export type { TopLevelEntity } from './src/model/TopLevelEntity';

// Null Objects
export { NoMapTypeEnumeration } from './src/model/MapTypeEnumeration';
export { NoTopLevelEntity } from './src/model/TopLevelEntity';

// Factories
export { newMetaEdEnvironment, newPluginEnvironment } from './src/MetaEdEnvironment';
export { newMetaEdPlugin } from './src/plugin/PluginTypes';

export { newAssociationProperty } from './src/model/property/AssociationProperty';
export { newBooleanProperty } from './src/model/property/BooleanProperty';
export { newChoiceProperty } from './src/model/property/ChoiceProperty';
export { newCommonProperty } from './src/model/property/CommonProperty';
export { newCurrencyProperty } from './src/model/property/CurrencyProperty';
export { newDateProperty } from './src/model/property/DateProperty';
export { newDecimalProperty } from './src/model/property/DecimalProperty';
export { newDescriptorProperty } from './src/model/property/DescriptorProperty';
export { newDomainEntityProperty } from './src/model/property/DomainEntityProperty';
export { newDurationProperty } from './src/model/property/DurationProperty';
export { newEnumerationProperty } from './src/model/property/EnumerationProperty';
export { newInlineCommonProperty } from './src/model/property/InlineCommonProperty';
export { newIntegerProperty } from './src/model/property/IntegerProperty';
export { newMergedProperty } from './src/model/property/MergedProperty';
export { newPercentProperty } from './src/model/property/PercentProperty';
export { newSchoolYearEnumerationProperty } from './src/model/property/SchoolYearEnumerationProperty';
export { newSharedDecimalProperty } from './src/model/property/SharedDecimalProperty';
export { newSharedIntegerProperty } from './src/model/property/SharedIntegerProperty';
export { newSharedShortProperty } from './src/model/property/SharedShortProperty';
export { newSharedStringProperty } from './src/model/property/SharedStringProperty';
export { newShortProperty } from './src/model/property/ShortProperty';
export { newSimpleProperty } from './src/model/property/SimpleProperty';
export { newStringProperty } from './src/model/property/StringProperty';
export { newTimeProperty } from './src/model/property/TimeProperty';
export { newYearProperty } from './src/model/property/YearProperty';

export { newAbstractEntity } from './src/model/AbstractEntity';
export { newAssociation } from './src/model/Association';
export { newAssociationExtension } from './src/model/AssociationExtension';
export { newAssociationSubclass } from './src/model/AssociationSubclass';
export { newChoice } from './src/model/Choice';
export { newCommon } from './src/model/Common';
export { newCommonExtension } from './src/model/CommonExtension';
export { newDecimalType } from './src/model/DecimalType';
export { newDescriptor } from './src/model/Descriptor';
export { newDomain } from './src/model/Domain';
export { newDomainEntity } from './src/model/DomainEntity';
export { newDomainEntityExtension } from './src/model/DomainEntityExtension';
export { newDomainEntitySubclass } from './src/model/DomainEntitySubclass';
export { newDomainItem } from './src/model/DomainItem';
export { newEnumeration } from './src/model/Enumeration';
export { newEnumerationItem } from './src/model/EnumerationItem';
export { newInlineCommon } from './src/model/Common';
export { newIntegerType } from './src/model/IntegerType';
export { newInterchange } from './src/model/Interchange';
export { newInterchangeExtension } from './src/model/InterchangeExtension';
export { newInterchangeItem } from './src/model/InterchangeItem';
export { newMapTypeEnumeration } from './src/model/MapTypeEnumeration';
export { newNamespaceInfo } from './src/model/NamespaceInfo';
export { newSchoolYearEnumeration } from './src/model/SchoolYearEnumeration';
export { newSharedDecimal } from './src/model/SharedDecimal';
export { newSharedInteger } from './src/model/SharedInteger';
export { newSharedString } from './src/model/SharedString';
export { newSourceMap } from './src/model/SourceMap';
export { newStringType } from './src/model/StringType';
export { newSubdomain } from './src/model/Subdomain';
export { asTopLevelEntity, newTopLevelEntity } from './src/model/TopLevelEntity';
