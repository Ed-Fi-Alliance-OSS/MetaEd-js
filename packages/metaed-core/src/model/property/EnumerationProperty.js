// @flow
import { ReferentialProperty, ReferentialPropertySourceMap, newReferentialProperty } from './ReferentialProperty';
import type { EntityPropertySourceMap, EntityProperty } from './EntityProperty';

export class EnumerationPropertySourceMap extends ReferentialPropertySourceMap {}

export class EnumerationProperty extends ReferentialProperty {
  sourceMap: EntityPropertySourceMap | ReferentialPropertySourceMap | EnumerationPropertySourceMap;
}

export function newEnumerationProperty(): EnumerationProperty {
  return Object.assign(new EnumerationProperty(), newReferentialProperty(), {
    type: 'enumeration',
    sourceMap: new EnumerationPropertySourceMap(),
  });
}

export const asEnumerationProperty = (x: EntityProperty): EnumerationProperty => ((x: any): EnumerationProperty);
