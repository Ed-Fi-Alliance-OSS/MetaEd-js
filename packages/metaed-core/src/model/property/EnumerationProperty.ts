// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import { ReferentialProperty, ReferentialPropertySourceMap } from './ReferentialProperty';
import { newReferentialProperty, newReferentialPropertySourceMap } from './ReferentialProperty';
import { EntityProperty } from './EntityProperty';

/**
 *
 */
export type EnumerationPropertySourceMap = ReferentialPropertySourceMap;

/**
 *
 */
export function newEnumerationPropertySourceMap(): EnumerationPropertySourceMap {
  return {
    ...newReferentialPropertySourceMap(),
  };
}

export interface EnumerationProperty extends ReferentialProperty {
  sourceMap: EnumerationPropertySourceMap;
}

/**
 *
 */
export function newEnumerationProperty(): EnumerationProperty {
  return {
    ...newReferentialProperty(),
    type: 'enumeration',
    typeHumanizedName: 'Enumeration Property',
    sourceMap: newEnumerationPropertySourceMap(),
  };
}

/**
 *
 */
export const asEnumerationProperty = (x: EntityProperty): EnumerationProperty => x as EnumerationProperty;
