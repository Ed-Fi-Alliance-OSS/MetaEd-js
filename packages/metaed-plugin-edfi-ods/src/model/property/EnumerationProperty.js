// @flow
import type { MetaEdEnvironment, EnhancerResult, EnumerationProperty } from 'metaed-core';
import { normalizeEnumerationSuffix, prependWithContextToMetaEdName } from '../../shared/Utility';
import type { EntityPropertyEdfiOds } from './EntityProperty';

export type EnumerationPropertyEdfiOds = EntityPropertyEdfiOds & {
  ods_TypeifiedBaseName: string,
};

// Enhancer for object setup
const enhancerName: string = 'EnumerationPropertySetupEnhancer';

export function odsTypeifiedBaseName(property: EnumerationProperty) {
  return normalizeEnumerationSuffix(property.metaEdName);
}

export function odsEnumerationName(property: EnumerationProperty) {
  return normalizeEnumerationSuffix(prependWithContextToMetaEdName(property.metaEdName, property.withContext));
}

export function addEnumerationPropertyEdfiOdsTo(property: EnumerationProperty) {
  if (property.data.edfiOds == null) property.data.edfiOds = {};

  Object.assign(property.data.edfiOds, {
    ods_TypeifiedBaseName: odsTypeifiedBaseName(property),
    ods_Name: odsEnumerationName(property),
  });
}

export function enhance(metaEd: MetaEdEnvironment): EnhancerResult {
  metaEd.propertyIndex.enumeration.forEach((property: EnumerationProperty) => {
    addEnumerationPropertyEdfiOdsTo(property);
  });

  return {
    enhancerName,
    success: true,
  };
}
