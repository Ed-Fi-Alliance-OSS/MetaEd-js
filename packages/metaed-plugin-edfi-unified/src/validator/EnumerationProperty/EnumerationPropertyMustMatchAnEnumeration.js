// @flow
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.enumeration.forEach(enumerationProperty => {
    if (metaEd.entity.enumeration.has(enumerationProperty.metaEdName)) return;
    failures.push({
      validatorName: 'EnumerationPropertyMustMatchAEnumeration',
      category: 'error',
      message: `Enumeration property ${enumerationProperty.metaEdName} does not match any declared enumeration.`,
      sourceMap: enumerationProperty.sourceMap.type,
      fileMap: null,
    });
  });
  return failures;
}
