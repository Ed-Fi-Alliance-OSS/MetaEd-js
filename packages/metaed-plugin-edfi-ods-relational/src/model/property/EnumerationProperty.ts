import { MetaEdEnvironment, EnhancerResult, EnumerationProperty } from '@edfi/metaed-core';
import { normalizeEnumerationSuffix } from '@edfi/metaed-core';
import { prependRoleNameToMetaEdName } from '../../shared/Utility';
import { EntityPropertyEdfiOds } from './EntityProperty';

export type EnumerationPropertyEdfiOds = EntityPropertyEdfiOds & {
  odsTypeifiedBaseName: string;
};

// Enhancer for object setup
const enhancerName = 'EnumerationPropertySetupEnhancer';

export function odsTypeifiedBaseName(property: EnumerationProperty) {
  return normalizeEnumerationSuffix(property.metaEdName);
}

export function odsEnumerationName(property: EnumerationProperty) {
  return normalizeEnumerationSuffix(prependRoleNameToMetaEdName(property.metaEdName, property.roleName));
}

export function addEnumerationPropertyEdfiOdsTo(property: EnumerationProperty) {
  if (property.data.edfiOdsRelational == null) property.data.edfiOdsRelational = {};

  Object.assign(property.data.edfiOdsRelational, {
    odsTypeifiedBaseName: odsTypeifiedBaseName(property),
    odsName: odsEnumerationName(property),
  });
}

export function enhance(metaEd: MetaEdEnvironment): EnhancerResult {
  metaEd.propertyIndex.enumeration.forEach((property: EnumerationProperty) => {
    addEnumerationPropertyEdfiOdsTo(property);
  });

  return {
    enhancerName,
    success: true,
  };
}
