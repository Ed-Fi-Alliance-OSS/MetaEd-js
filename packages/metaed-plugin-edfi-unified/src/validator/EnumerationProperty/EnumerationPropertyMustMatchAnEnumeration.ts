import { MetaEdEnvironment, ValidationFailure, ModelBase } from 'metaed-core';
import { getEntityFromNamespaceChain } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.propertyIndex.enumeration.forEach(property => {
    const referencedEntity: ModelBase | null = getEntityFromNamespaceChain(
      property.metaEdName,
      property.referencedNamespaceName,
      property.namespace,
      'enumeration',
    );

    if (referencedEntity == null) {
      failures.push({
        validatorName: 'EnumerationPropertyMustMatchAEnumeration',
        category: 'error',
        message: `Enumeration property '${property.metaEdName}' does not match any declared Enumeration in namespace ${
          property.referencedNamespaceName
        }.`,
        sourceMap: property.sourceMap.metaEdName,
        fileMap: null,
      });
    }
  });
  return failures;
}
