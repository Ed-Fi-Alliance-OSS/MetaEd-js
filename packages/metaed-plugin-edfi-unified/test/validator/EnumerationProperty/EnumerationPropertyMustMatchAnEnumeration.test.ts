import {
  newMetaEdEnvironment,
  MetaEdTextBuilder,
  EnumerationBuilder,
  DomainEntityBuilder,
  NamespaceBuilder,
} from '@edfi/metaed-core';
import { MetaEdEnvironment, ValidationFailure } from '@edfi/metaed-core';
import { validate } from '../../../src/validator/EnumerationProperty/EnumerationPropertyMustMatchAnEnumeration';

describe('when enumeration property has valid identifier', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const entityName = 'EntityName1';
  let failures: ValidationFailure[];
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartEnumeration(entityName)
      .withDocumentation('EntityDocumentation')
      .withEndEnumeration()

      .withStartDomainEntity('EntityName2')
      .withDocumentation('EntityDocumentation')
      .withStringIdentity('PropertyName1', 'PropertyDocumentation', '100')
      .withEnumerationProperty(entityName, 'PropertyDocumentation', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new EnumerationBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build one enumeration', (): void => {
    expect(coreNamespace.entity.enumeration.size).toBe(1);
  });

  it('should build one domain entity', (): void => {
    expect(coreNamespace.entity.domainEntity.size).toBe(1);
  });

  it('should have no validation failures', (): void => {
    expect(failures).toHaveLength(0);
  });
});

describe('when enumeration property has invalid identifier', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: ValidationFailure[];
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity('EntityName1')
      .withDocumentation('EntityDocumentation')
      .withStringIdentity('PropertyName1', 'PropertyDocumentation', '100')
      .withEnumerationProperty('PropertyName2', 'PropertyDocumentation', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build one domain entity', (): void => {
    expect(coreNamespace.entity.domainEntity.size).toBe(1);
  });

  it('should have validation failures', (): void => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('EnumerationPropertyMustMatchAEnumeration');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot();
    expect(failures[0].sourceMap).toMatchSnapshot();
  });
});
