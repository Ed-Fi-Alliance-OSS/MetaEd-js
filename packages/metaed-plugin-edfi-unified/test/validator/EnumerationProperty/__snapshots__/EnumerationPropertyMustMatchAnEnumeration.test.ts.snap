// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when enumeration property has invalid identifier should have validation failures 1`] = `"Enumeration property PropertyName2 does not match any declared enumeration."`;

exports[`when enumeration property has invalid identifier should have validation failures 2`] = `
Object {
  "column": 16,
  "line": 10,
  "tokenText": "PropertyName2",
}
`;
